hugo/content/tutorials/generation_in_the_web.md

```diff
@@ -5,391 +5,172 @@ weight: 7
 }
 ```
 
-And now our implementation features a custom command handler that takes a MiniLogo program, and returns a generated result from that program's AST. To get these changes into the language server itself, you'll want to rebuild & bundle everything once more. If you recall the command from the last tutorial, we can do this via `build:web`.
-
-```bash
-npm run build:web
-```
+Now this works, but when do we receive notifications, and how often? Well a good thing you asked, because if you started this up and began editing your program, you would be receiving a notification for every single change! Including whitespace changes. Now that's probably not what we're looking for, but the content is correct, we just want to slow it down a bit. We can do this by setting a timeout and a semaphore to prevent multiple notifications from being processed at once.
 
-## Importing the Generator
-
-Now, if you've been following along with our prior tutorials, you should have a **src/static/** folder already setup with an HTML and JS file, plus an updated language server bundle. We can now go into the HTML file, and make a couple changes to our HTML file to get things ready to work with our new changes.
-
-- add a canvas
-- add a button to trigger updating the canvas
-
-You should also replace the previous HTML & CSS files with the following contents. The HTML updates add Monaco, a Canvas, and a build button. The CSS styles these new additions so that they're properly aligned.
-
-{{< tabs "new-html-css" >}}
-{{< tab "HTML" >}}
-
-```html
-<!DOCTYPE html>
-<html>
-    <head>
-        <meta charset='utf-8'>
-        <!-- Page & Monaco styling -->
-        <link href="styles.css" rel="stylesheet"/>
-        <title>MiniLogo in Langium</title>
-    </head>
-    <body>
-        <h1>MiniLogo in Langium</h1>
-
-        <!-- Use a wrapper to display Monaco + Canvas side-by-side -->
-        <div id="page-wrapper">
-            <!-- Monaco half -->
-            <div class="half">
-                <div class="wrapper">
-                    <div id="monaco-editor-root"></div>
-                </div>
-            </div>
-            <!-- Canvas half -->
-            <div class="half">
-                <canvas id='minilogo-canvas' width=500 height=600></canvas>
-            </div>
-        </div>
-
-        <!-- Add a button to update our canvas, will invoke a globally accessible function -->
-        <div>
-            <input class="build" type="button" value="Update Canvas" onclick="window.generateAndDisplay()">
-        </div>
-
-        <br/>
-        <footer>
-            <br/>
-            <p style="font-style:italic">Powered by</p>
-            <img width="125" src="https://langium.org/assets/langium_logo_w_nib.svg" alt="Langium">
-        </footer>
-        <!-- Monaco Configuration -->
-        <script type="module" src="setup.js"></script>
-    </body>
-</html>
-```
+```ts
+let running = false;
+let timeout: NodeJS.Timeout | null = null;
```

Comment:
`NodeJS.Timeout` on the client side?

Reply:
It's used as part of the throttle mechanic, so no requests are dispatched until we have X amount of time delay following (200ms I think is the current value).

Reply:
I was just wondering if the code above is browser-compatible.

Reply:
Good point, it should be since the types are erased, but yeah that's the wrong type for a browser-based context. It should be a number instead (positive int to be specific).